src/stream/src/executor/managed_state/join/mod.rs

```diff
@@ -151,149 +156,176 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
         let state = JoinEntryState::new(
             keyspace,
-            self.data_types.clone(),
-            self.pk_data_types.clone(),
+            table_info.data_types.clone(),
+            table_info.pk_data_types.clone(),
         );
-        self.inner.put(key.clone(), state);
-        Ok(())
+        Ok(state)
     }
 
-    /// Get or create a [`JoinEntryState`] without cached state. Should only be called if the key
-    /// does not exist in memory or remote storage.
-    pub async fn get_or_init_without_cache(&mut self, key: &K) -> RwResult<&mut JoinEntryState<S>> {
-        // TODO: we should probably implement a entry function for `LruCache`
-        let contains = self.inner.contains(key);
-        if contains {
-            Ok(self.inner.get_mut(key).unwrap())
-        } else {
-            self.init_without_cache(key).await?;
-            Ok(self.inner.get_mut(key).unwrap())
+    pub fn pop_cached(&mut self, key: &K) -> Option<JoinEntryState<S>> {
+        self.inner.pop(key)
+    }
+
+    pub fn push(&mut self, key: K, state: JoinEntryState<S>) {
+        self.inner.push(key, state);
+    }
+
+    // Iter keys on the
+    pub(crate) fn iter_keys<'a>(
+        &'a mut self,
+        side_match: &'a mut Self,
+        keys: impl Iterator<Item = &'a K>,
+    ) -> IterJoinEntriesMut<'a, K, S> {
+        IterJoinEntriesMut::<'a>::new(side_match, self, keys)
+    }
+}
+
+type IterType<K, S> = (K, Option<JoinEntryState<S>>, JoinEntryState<S>);
+
+/// Iterates
+pub(crate) struct IterJoinEntriesMut<'a, K: HashKey, S: StateStore> {
+    side_match: &'a mut JoinHashMap<K, S>,
+    side_update: &'a mut JoinHashMap<K, S>,
+    last_item: Option<IterType<K, S>>,
+    queue: tokio::sync::mpsc::UnboundedReceiver<IterType<K, S>>, /* TODO: change this to a
+                                                                  * reasonably
+                                                                  * sized bounded queue... */
+}
+
+impl<'a, K: HashKey, S: StateStore> IterJoinEntriesMut<'a, K, S> {
+    pub(crate) fn new(
+        side_match: &'a mut JoinHashMap<K, S>,
+        side_update: &'a mut JoinHashMap<K, S>,
+        keys: impl Iterator<Item = &'a K>,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel::<IterType<K, S>>();
+        let new_iter = Self {
+            side_match,
+            side_update,
+            last_item: None,
+            queue: rx,
+        };
+
+        for key in keys {
+            let update_entry = new_iter.side_update.pop_cached(key);
+            let match_entry = if !key.has_null() {
+                new_iter.side_match.pop_cached(key)
+            } else {
+                None
+            };
+            let key = key.clone();
+            let match_table_info = new_iter.side_match.table_info.clone();
+            let update_table_info = new_iter.side_update.table_info.clone();
+            let tx = tx.clone();
+
+            tokio::spawn(async move {
```

Comment:
Maybe it would be better to avoid spawning a lot of tasks by using `stream::iter(futures)` like in `hash_agg`.

Reply:
Hmm, my understanding is that tasks are the most lightweight unit of concurrency, and that `stream::iter` is just a way of wrapping tasks and queue in a simpler abstraction.

Reply:
A task is a unit scheduled by tokio, which means they can be scheduled to different threads while `stream::iter` make all futures in 1 task.  The parallelism in our system should be actor level -- better to have one actor working on one thread.  Besides, it might be bad to let tokio schedule key number of tasks.